### PR TITLE
WIP: Cython cleanup

### PR DESCRIPTION
**Pull request by KaiSzuttor**
Move ```pxd``` files to a sub-directory to distinguish between interface and core.

### CODE REVIEW

**Comment by fweik** [comment]
I think you should document the idea somewhere: Our idea was to separate the pure core bindings
from the interface code. Also probably there needs to be some cmake support for this, so that the
dependencies between the python modules can be expressed. So I think there should be the concept
of a cython module in cmake that are cmake targets.

**Comment by KaiSzuttor** [comment]
the cmake code for cython is currently kind of messy... there we would have to start probably, removing wildcards etc.

**Comment by fweik** [comment]
Yes, better cmake may very well be a side effect of this ^^

**Comment by KaiSzuttor** [comment]
we probably want sth like that...
https://github.com/scikit-build/scikit-build/blob/master/skbuild/resources/cmake/UseCython.cmake

**Comment by fweik** [comment]
I think it's sufficient to just put the cmake code we have into a function, similar to e.g. the `unit_test` function, and clean it up a bit.

**Comment by RudolfWeeber** [comment]
I don't thik moving part of the pxd files simplifies matters.
Particularly, when there are additional pxd files containing cdef inline functions. Then, there are two places to look for stuff.

I'd prefer keeping the pxd files with the pyx files.

cdef extern declarations are also allowed in .pyx files. The only reason for having pxd files is when two or more pyx files use them.
Probably, we could avoid most pxd files entirely and put the cdef externs close to where they are used.

Can we discuss this in an ES meeting?

**Comment by fweik** [comment]
@RudolfWeeber maybe there is a misunderstanding: This isn't about separating the interface implementation from their headers, but separating the core bindings from the interface implementation. There is no pxd/pyx split there (because there are no implementations), and the plan is to have a one-to-one mapping from the core headers to pxd files. This seems clearer to me then the current state where everything is mixed.

**Comment by KaiSzuttor** [comment]
> Probably, we could avoid most pxd files entirely and put the cdef externs close to where they are used.

This is a bad idea. It would be really confusing if some extern declarations occur within pyx files and some in pxd files.

**Comment by jngrad** [comment]
@KaiSzuttor are you sure about the merge? It re-introduced formatting issues from #2974, even though I took care not to modify code already touched by this PR. If I remember correctly, when test-merging this PR into #2974, the only conflict was at line 622 of electrostatics.pyx.

**Comment by KaiSzuttor** [comment]
@jngrad I remerged...

**Comment by KaiSzuttor** [comment]
@fweik i think this PR would improve the interface code quite a lot. could you have a look?
@RudolfWeeber can you please explain your concerns again?

**Comment by fweik** [comment]
This is fine with me, but I think it is only cosmetic. There needs to be change in the core first, e.g. separating the callback stuff from the synchronous actual core. It's not clear to me why you started here.

**Comment by KaiSzuttor** [comment]
it doesn't really matter where to start. I think it's easier to see flaws in the core now. Therefore, I think it's more than cosmetics...

**Comment by fweik** [comment]
Fair enough.

File: doc/sphinx/conf.py.in [hunk 1]
```diff
@@ -86,7 +86,7 @@ language = None
 
 # List of patterns, relative to source directory, that match files and
 # directories to ignore when looking for source files.
-exclude_patterns = ["espressomd.gen_code_info*", "espressomd.gen_pxiconfig*"]
+exclude_patterns = ["espressomd.gen_code_info*", "espressomd.gen_pxiconfig*", "espressomd.core.rst"]
```
Comment (fweik) [inline doc/sphinx/conf.py.in:89]: Why `.rst`?

Comment (KaiSzuttor) [reply doc/sphinx/conf.py.in:89]: because it's the name of the file we want to exclude

Comment (KaiSzuttor) [reply doc/sphinx/conf.py.in:89]: actually we probably don't need those exclude patterns at all if they are excluded in the ```CMakeLists.txt``` file

Comment (fweik) [reply doc/sphinx/conf.py.in:89]: I can't find it, did you check that in?

Comment (KaiSzuttor) [reply doc/sphinx/conf.py.in:89]: it gets generated, but probably not if the tool is called correctly

**Comment by fweik** [comment]
Why did you abandon this?

**Comment by KaiSzuttor** [comment]
i currently don't have time for it and it's not too important for the 4.1 release

**Pull request closed without merging**